corona-warn-app/cwa-app-android
# Overhaul of DiagnosisKey related Download and Caching Mechanisms (EXPOSUREAPP-2469)

**This PR is an overhaul of the existing logic for downloading and caching diagnosis keyfiles.**

* Supports multiple countries (interop).
* Removes any singleton use in this scope and replaces it with dependency injection.
* Adds tests for all classes in this scope (~100 new tests).
* Migrates old cache files to the new logic to preserve bandwitdth.
* Download API, key fetching and caching is now strongly typed (no longer using raw strings) to possibly catch errors at compile time.
* The download API is now encoded in a `DownloadAPIV1` and no longer using string constants. If there is a V2 api, we will make a new interface.
* The KeyCacheRespoitory` hides all file/storage details and only provides direct access to paths on demand (to prevent hardcoding paths that may change, i.e. move to sdcard). It also uses an unencrypted database for storing the keyfile metadata to offer less bug attack surface encryption issues (i.e. error 9002). We don't need encryption for information that is public anyways all files + meta data comes from a public CDN that everyone has access to.
* The `KeyFileDownloader` is now just responsible for determining the missing days and downloading them, i.e. syncing our cache with the server. The internal "fetch last 3 hours" is no longer "TODO remove" and has been rewritten to potentially support future hour based key file download.

**TODO:**
* ~~Complete tests for `KeyFileDownloader`~~
* Wait for architecture decision if we should continue if any file download fails (which it now does due to higher chance for download failures when downloading countries*days files).
* ~~Replace `FileStorageHelper` with `DeviceStorage`~~

**How to test**
* Use the app and trigger key downloads in various situations
* Enter the test menu and trigger repeated downloads
* Open logcat and filter for `Download`
* Install an old version (dev) fetch keys, upgrade to this PR, install, fetch keys, check our apps cache folder, the old keys should be gone and the log should show that we migrated the old files, and we didn't download the whole again (it just get's the files checksum from the headers and compares it with our old keys).
